File: packages/shared/src/services/ingest-router.ts
/**
 * ingest-router — YAML-driven routing for uploaded files (CS3.11).
 *
 * TypeScript mirror of `scripts/lib/ingest_router.py` (CS3.12). Both sides
 * consume the same `config/ingest-routes.yaml` so the upload endpoint
 * (CS3.4) and the Python sidecar never drift on which filename maps to
 * which pipeline + parser.
 *
 * Matching semantics follow Python's `fnmatch` (case-insensitive):
 *   `*` — any characters
 *   `?` — single character
 *   `[abc]` — character class
 *   everything else literal
 *
 * Also handles dispatch: given a `source_type`, build the compose-network
 * URL for the matching sidecar (`http://{source}-ingest:8080`) and POST
 * a `/process` request with the bearer-secret header. The response is
 * validated against `SidecarProcessResponseSchema` from `@open-brain/shared`.
 *
 * Moved from `packages/core-api/src/services/ingest-router.ts` to shared
 * (CS3.5) so both core-api routes and workers jobs can import a single
 * implementation and cannot drift.
 */

import { readFileSync } from 'node:fs'
import { basename, join } from 'node:path'
import yaml from 'js-yaml'
import { HttpError } from '../utils/fetch-helpers.js'
import { createLogger } from '../lib/logger.js'
import {
  SidecarProcessResponseSchema,
  type SidecarProcessResponse,
} from '../schema/ingest.js'

const logger = createLogger('ingest-router')

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

/**
 * A single routing rule as it appears in the YAML (flattened with the
 * parent source_type + derived pipeline filename for convenience).
 */
export interface IngestRoute {
  source_type: string
  /** Conventional Python pipeline filename (`{source}-pipeline.py`). */
  pipeline: string
  parser: string
  pattern: string
  header_sniff: string | null
}

interface RawRule {
  pattern?: unknown
  parser?: unknown
  header_sniff?: unknown
}

interface RawRoutesYaml {
  routes?: Record<string, unknown>
}

// ---------------------------------------------------------------------------
// YAML load + cache
// ---------------------------------------------------------------------------

let _cache: IngestRoute[] | null = null
let _cacheSource: string | null = null

const DEFAULT_CONTAINER_PATH = '/app/config/ingest-routes.yaml'

function resolveRoutesPath(explicit?: string): string {
  if (explicit) return explicit
  const envPath = process.env.INGEST_ROUTES_PATH
  if (envPath) return envPath
  // Prefer container default when present; fall back to repo-relative path.
  try {
    readFileSync(DEFAULT_CONTAINER_PATH, 'utf-8')
    return DEFAULT_CONTAINER_PATH
  } catch {
    return join(process.cwd(), 'config', 'ingest-routes.yaml')
  }
}

/**
 * Load (and memoize) the routes from `config/ingest-routes.yaml`.
 *
 * Returns a flat list of `IngestRoute` — each rule carries its parent
 * `source_type` and derived `pipeline` filename. Source-type order and
 * within-source rule order match the YAML declaration order; callers that
 * match by filename should iterate in order (first match wins), matching
 * the Python side.
 *
 * @param explicitPath Optional override — primarily for tests.
 */
export function loadRoutes(explicitPath?: string): IngestRoute[] {
  const path = resolveRoutesPath(explicitPath)

  if (_cache !== null && _cacheSource === path) {
    return _cache
  }

  let raw: string
  try {
    raw = readFileSync(path, 'utf-8')
  } catch (err) {
    logger.error({ path, err: err instanceof Error ? err.message : String(err) }, 'Failed to read ingest-routes.yaml')
    throw new Error(`ingest-routes.yaml not found at ${path}`)
  }

  const parsed = yaml.load(raw) as RawRoutesYaml | null | undefined
  const routes: IngestRoute[] = []

  if (!parsed || typeof parsed !== 'object') {
    logger.warn({ path }, 'ingest-routes.yaml is empty or not a mapping — using empty route set')
    _cache = routes
    _cacheSource = path
    return routes
  }

  const rawRoutes = parsed.routes
  if (rawRoutes && typeof rawRoutes === 'object') {
    // Preserve declaration order from the YAML (js-yaml returns insertion-ordered objects).
    for (const [sourceType, rules] of Object.entries(rawRoutes)) {
      if (!Array.isArray(rules)) continue
      for (const rule of rules as RawRule[]) {
        if (!rule || typeof rule !== 'object') continue
        const pattern = typeof rule.pattern === 'string' ? rule.pattern : null
        const parser = typeof rule.parser === 'string' ? rule.parser : null
        if (!pattern || !parser) continue
        const headerSniff = typeof rule.header_sniff === 'string' ? rule.header_sniff : null
        routes.push({
          source_type: sourceType,
          pipeline: `${sourceType}-pipeline.py`,
          parser,
          pattern,
          header_sniff: headerSniff,
        })
      }
    }
  }

  _cache = routes
  _cacheSource = path
  logger.debug({ path, routeCount: routes.length }, 'Loaded ingest routes')
  return routes
}

/** Reset the module cache — exported for tests. */
export function _clearIngestRoutesCache(): void {
  _cache = null
  _cacheSource = null
}

// ---------------------------------------------------------------------------
// Lookup helpers
// ---------------------------------------------------------------------------

/**
 * Return all routes for a given source_type (in declaration order), or
 * `null` if the source_type is unknown.
 */
export function routeForSourceType(sourceType: string, explicitPath?: string): IngestRoute | null {
  if (!sourceType) return null
  const routes = loadRoutes(explicitPath)
  const match = routes.find((r) => r.source_type === sourceType)
  return match ?? null
}

/**
 * Return the first route whose pattern matches `filename`, or `null`.
 * Matching is case-insensitive and uses fnmatch-style globs (see module
 * header). Matches the Python `route_for_filename` behaviour so both
 * sides agree on dispatch for every filename.
 */
export function routeForFilename(filename: string, explicitPath?: string): IngestRoute | null {
  if (!filename) return null
  const base = basename(filename).toLowerCase()
  const routes = loadRoutes(explicitPath)
  for (const route of routes) {
    if (fnmatchCaseInsensitive(base, route.pattern)) {
      return route
    }
  }
  return null
}

// ---------------------------------------------------------------------------
// fnmatch-compatible glob matcher
// ---------------------------------------------------------------------------

/**
 * Translate an fnmatch-style glob into a regex source string. Mirrors
 * CPython's `fnmatch.translate` closely enough for the patterns used in
 * `ingest-routes.yaml` (which are simple `*` / `?` / literal globs with
 * no character classes in the seeded file, but we support `[...]` too
 * for forward compatibility).
 */
function translateGlob(pattern: string): string {
  let i = 0
  const n = pattern.length
  let out = ''
  while (i < n) {
    const c = pattern[i]
    i++
    if (c === '*') {
      out += '.*'
    } else if (c === '?') {
      out += '.'
    } else if (c === '[') {
      // Find the closing bracket.
      let j = i
      if (j < n && pattern[j] === '!') j++
      if (j < n && pattern[j] === ']') j++
      while (j < n && pattern[j] !== ']') j++
      if (j >= n) {
        // No closing bracket — treat '[' as literal.
        out += '\\['
      } else {
        let stuff = pattern.slice(i, j).replace(/\\/g, '\\\\')
        if (stuff.startsWith('!')) stuff = '^' + stuff.slice(1)
        out += `[${stuff}]`
        i = j + 1
      }
    } else {
      // Escape regex metacharacters.
      out += c.replace(/[-/\\^$+?.()|[\]{}]/g, '\\$&')
    }
  }
  return `^${out}$`
}

function fnmatchCaseInsensitive(filename: string, pattern: string): boolean {
  const re = new RegExp(translateGlob(pattern.toLowerCase()))
  return re.test(filename.toLowerCase())
}

// ---------------------------------------------------------------------------
// Sidecar dispatch
// ---------------------------------------------------------------------------

/**
 * Return the compose-network URL for the sidecar that handles a given
 * source_type. Defaults to `http://{source}-ingest:8080` (matches the
 * compose service names `financial-ingest` / `utility-ingest`). Can be
 * overridden per source_type via env `INGEST_SIDECAR_URL_<SOURCE_UPPER>`
 * — primarily for local debugging.
 */
export function sidecarUrlForSourceType(sourceType: string): string {
  const envKey = `INGEST_SIDECAR_URL_${sourceType.toUpperCase()}`
  const override = process.env[envKey]
  if (override && override.length > 0) return override.replace(/\/+$/, '')
  return `http://${sourceType}-ingest:8080`
}

export interface DispatchToSidecarInput {
  sourceType: string
  fileId: string
  filePath: string
  /** Shared secret; callers should read from `process.env.INGEST_TRIGGER_SECRET`. */
  secret: string
  /** Default 300_000 ms (5 min) to match the sidecar subprocess timeout. */
  timeoutMs?: number
}

/**
 * POST `/process` to the sidecar container for the given source_type and
 * parse the response via `SidecarProcessResponseSchema`.
 *
 * Throws `HttpError` (from `@open-brain/shared`) with status + body on
 * non-2xx responses, or a plain `Error` on timeout / network failure.
 * Callers (e.g. the CS3.5 BullMQ worker) are responsible for retry
 * policy; this function does a single attempt.
 */
export async function dispatchToSidecar(input: DispatchToSidecarInput): Promise<SidecarProcessResponse> {
  const { sourceType, fileId, filePath, secret } = input
  const timeoutMs = input.timeoutMs ?? 300_000

  const url = `${sidecarUrlForSourceType(sourceType)}/process`
  const controller = new AbortController()
  const timer = setTimeout(() => controller.abort(), timeoutMs)

  try {
    const res = await fetch(url, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'X-Open-Brain-Caller': 'ingest',
        Authorization: `Bearer ${secret}`,
      },
      body: JSON.stringify({ file_path: filePath, file_id: fileId }),
      signal: controller.signal,
    })

    if (!res.ok) {
      const body = await res.text().catch(() => '')
      throw new HttpError(res.status, body, `sidecar ${sourceType} /process`)
    }

    const json = (await res.json()) as unknown
    return SidecarProcessResponseSchema.parse(json)
  } catch (err) {
    if (err instanceof HttpError) throw err
    if (err instanceof Error && err.name === 'AbortError') {
      throw new Error(`sidecar ${sourceType} /process timed out after ${timeoutMs}ms`)
    }
    throw err
  } finally {
    clearTimeout(timer)
  }
}

File: scripts/lib/ingest_router.py
"""YAML-driven ingest router for the Python sidecar (CS3.12).

Reads ``config/ingest-routes.yaml`` — the shared source-of-truth consumed by
both the TypeScript upload endpoint (``packages/core-api/src/services/
ingest-router.ts`` — CS3.11) and this Python helper — so the TS and Python
sides never drift on which filename maps to which pipeline + parser.

YAML shape (as seeded in CS2.9)::

    routes:
      financial:
        - pattern: "activity.csv"
          parser:  amex
        - pattern: "Download*.csv"
          parser:  paypal
          header_sniff: '"Balance Impact"'
      utility:
        - pattern: "*gas*.pdf"
          parser:  gas_bill

Each ``source_type`` (``financial`` / ``utility`` / future) maps to an
ordered list of ``{pattern, parser, header_sniff?}`` rules. ``pattern`` is
a glob (matched with ``fnmatch`` against the basename, case-insensitive),
not a regex. First match wins. ``header_sniff`` is advisory here — the
pipeline scripts themselves do the actual CSV header probing to
disambiguate generic filenames (e.g. PayPal "Download.csv").

The pipeline script filename is derived from the source_type by
convention: ``{source_type}-pipeline.py`` (sits next to this module in
``scripts/``). This keeps the YAML declarative (no script paths baked in)
while letting ``trigger_server.py`` resolve a source_type → absolute
path at runtime.

Public API
----------

* ``load_routes(config_path=None)`` — parse + cache the YAML.
* ``route_for_source(source_type)`` — return the routing block for a
  source_type (list of rules).
* ``route_for_filename(filename)`` — return the first rule matching a
  filename, with ``source_type`` / ``pipeline`` / ``parser`` populated.
* ``script_for_source(source_type)`` — return the pipeline script name
  (e.g. ``financial-pipeline.py``). This is the hook
  ``docker/ingest-sidecar/trigger_server.py`` imports.
* ``resolve_command_args(route, file_path, extra_args=None)`` — build the
  subprocess argv for running the pipeline against a single file.
* ``list_source_types()`` — sorted source_type list (for diagnostics).

All functions are safe to call from any process; the YAML is loaded once
per process and memoized in a module-level cache. Call
``_clear_cache()`` (private) from tests that need a fresh load.
"""

from __future__ import annotations

import fnmatch
import json
import os
import sys
from pathlib import Path
from typing import Any

try:
    import yaml  # PyYAML
except ImportError as _yaml_exc:  # pragma: no cover — hard dep
    yaml = None  # type: ignore[assignment]
    _YAML_IMPORT_ERROR: Exception | None = _yaml_exc
else:
    _YAML_IMPORT_ERROR = None


# ---------------------------------------------------------------------------
# Defaults + module-level cache
# ---------------------------------------------------------------------------

DEFAULT_CONTAINER_PATH = "/app/config/ingest-routes.yaml"
DEFAULT_REPO_PATH = "config/ingest-routes.yaml"

_cache: dict[str, Any] | None = None
_cache_source: str | None = None


def _resolve_config_path(config_path: str | Path | None) -> Path:
    """Resolve the YAML path with env-var + fallback support.

    Precedence:
      1. Explicit ``config_path`` argument.
      2. ``$INGEST_ROUTES_PATH`` env var.
      3. ``/app/config/ingest-routes.yaml`` (container default).
      4. ``./config/ingest-routes.yaml`` relative to cwd (repo default).
    """
    if config_path is not None:
        return Path(config_path)
    env = os.environ.get("INGEST_ROUTES_PATH")
    if env:
        return Path(env)
    container = Path(DEFAULT_CONTAINER_PATH)
    if container.exists():
        return container
    return Path(DEFAULT_REPO_PATH)


def _clear_cache() -> None:
    """Reset the module cache — intended for tests."""
    global _cache, _cache_source
    _cache = None
    _cache_source = None


def load_routes(config_path: str | Path | None = None) -> dict[str, Any]:
    """Load + parse ``ingest-routes.yaml`` and cache the result.

    Returns the full parsed YAML document. Raises ``FileNotFoundError`` if
    the file is missing and ``RuntimeError`` if PyYAML isn't installed.
    """
    global _cache, _cache_source

    if yaml is None:  # pragma: no cover — hard dep
        raise RuntimeError(
            "PyYAML is required for ingest_router but is not installed. "
            f"Original import error: {_YAML_IMPORT_ERROR!r}"
        )

    path = _resolve_config_path(config_path)
    key = str(path.resolve()) if path.exists() else str(path)

    if _cache is not None and _cache_source == key:
        return _cache

    if not path.exists():
        raise FileNotFoundError(f"ingest-routes.yaml not found at {path!s}")

    with path.open("r", encoding="utf-8") as fh:
        parsed = yaml.safe_load(fh) or {}

    if not isinstance(parsed, dict):
        raise ValueError(
            f"ingest-routes.yaml must be a mapping at top level, got {type(parsed).__name__}"
        )

    _cache = parsed
    _cache_source = key
    return parsed


# ---------------------------------------------------------------------------
# Source-type / pipeline helpers
# ---------------------------------------------------------------------------


def list_source_types(config_path: str | Path | None = None) -> list[str]:
    """Return the sorted list of known source types."""
    routes = load_routes(config_path).get("routes") or {}
    if not isinstance(routes, dict):
        return []
    return sorted(routes.keys())


def route_for_source(
    source_type: str,
    config_path: str | Path | None = None,
) -> list[dict[str, Any]] | None:
    """Return the routing-rule list for ``source_type`` (or ``None``)."""
    routes = load_routes(config_path).get("routes") or {}
    if not isinstance(routes, dict):
        return None
    rules = routes.get(source_type)
    if rules is None:
        return None
    if not isinstance(rules, list):
        return None
    # Defensive copy so callers can't mutate the cache.
    return [dict(r) for r in rules if isinstance(r, dict)]


def script_for_source(
    source_type: str,
    config_path: str | Path | None = None,
) -> str | None:
    """Return the pipeline script filename for ``source_type``.

    By convention: ``{source_type}-pipeline.py``. Returns ``None`` for
    unknown source types. This is the function
    ``docker/ingest-sidecar/trigger_server.py`` imports to resolve a
    source_type into an executable script — keep the name stable.
    """
    if not source_type:
        return None
    # Only acknowledge source types that actually appear in the YAML so we
    # don't silently fabricate pipeline names for typos.
    known = list_source_types(config_path)
    if source_type not in known:
        return None
    return f"{source_type}-pipeline.py"


# ---------------------------------------------------------------------------
# Filename → route dispatch
# ---------------------------------------------------------------------------


def route_for_filename(
    filename: str,
    config_path: str | Path | None = None,
) -> dict[str, Any] | None:
    """Return the first route matching ``filename``, or ``None``.

    Matching:
      * Glob-style via ``fnmatch`` against the basename.
      * Case-insensitive — both the pattern and filename are lowercased.
      * Iterates source types in sorted order, then rules in declaration
        order within each source type. First hit wins.

    The returned dict is a flattened view of the rule plus its
    source_type and the derived pipeline script::

        {
          "source_type":  "financial",
          "pipeline":     "financial-pipeline.py",
          "parser":       "amex",
          "pattern":      "activity.csv",
          "header_sniff": None,  # or the literal string from YAML
        }
    """
    if not filename:
        return None

    basename = os.path.basename(filename).lower()
    routes = load_routes(config_path).get("routes") or {}
    if not isinstance(routes, dict):
        return None

    for source_type in sorted(routes.keys()):
        rules = routes.get(source_type) or []
        if not isinstance(rules, list):
            continue
        for rule in rules:
            if not isinstance(rule, dict):
                continue
            pattern = str(rule.get("pattern", ""))
            if not pattern:
                continue
            if fnmatch.fnmatchcase(basename, pattern.lower()):
                return {
                    "source_type": source_type,
                    "pipeline": f"{source_type}-pipeline.py",
                    "parser": rule.get("parser"),
                    "pattern": pattern,
                    "header_sniff": rule.get("header_sniff"),
                }
    return None


# ---------------------------------------------------------------------------
# Subprocess argv helper
# ---------------------------------------------------------------------------


def resolve_command_args(
    route: dict[str, Any],
    file_path: str,
    extra_args: list[str] | None = None,
) -> list[str]:
    """Build the subprocess argv for running a pipeline against one file.

    Mirrors the invocation pattern in ``trigger_server.run_pipeline`` so
    callers wanting to dispatch a single file (not an inbox sweep) get a
    consistent command shape::

        python scripts/<pipeline> --process-inbox --file <path> --json-output [extra...]

    ``sys.executable`` is used over the bare string ``"python"`` so the
    same invocation works inside the sidecar (python3.12) and on the
    host (wherever ``python`` resolves).
    """
    if not isinstance(route, dict):
        raise TypeError("route must be a dict returned by route_for_filename")
    pipeline = route.get("pipeline")
    if not pipeline:
        raise ValueError("route is missing 'pipeline'")

    argv: list[str] = [
        sys.executable or "python",
        f"scripts/{pipeline}",
        "--process-inbox",
        "--file",
        str(file_path),
        "--json-output",
    ]
    if extra_args:
        argv.extend(extra_args)
    return argv


# ---------------------------------------------------------------------------
# CLI entrypoint — ``python -m scripts.lib.ingest_router --list``
# ---------------------------------------------------------------------------


def _main(argv: list[str]) -> int:  # pragma: no cover — CLI shim
    if "--list" in argv or "-l" in argv:
        out = {
            "source_types": list_source_types(),
            "routes": load_routes().get("routes", {}),
        }
        print(json.dumps(out, indent=2, sort_keys=True))
        return 0
    if "--match" in argv:
        i = argv.index("--match")
        if i + 1 >= len(argv):
            print("usage: --match <filename>", file=sys.stderr)
            return 2
        match = route_for_filename(argv[i + 1])
        print(json.dumps(match, indent=2, sort_keys=True))
        return 0
    print(
        "ingest_router — YAML-driven routing helper\n"
        "usage:\n"
        "  python -m scripts.lib.ingest_router --list\n"
        "  python -m scripts.lib.ingest_router --match <filename>",
        file=sys.stderr,
    )
    return 2


if __name__ == "__main__":  # pragma: no cover
    raise SystemExit(_main(sys.argv[1:]))
